src/hll/sketch.rs

```diff
@@ -0,0 +1,422 @@
+//! - HIP accumulator state for accurate estimation
+//! - Out-of-order flag for merged/deserialized sketches
+//!
+//! The serialization format is compatible with Apache DataSketches implementations
+//! in Java and C++, enabling cross-platform sketch exchange.
+
+use std::hash::Hash;
+use std::io;
+
+use crate::hll::array4::Array4;
+use crate::hll::array6::Array6;
+use crate::hll::array8::Array8;
+use crate::hll::container::Container;
+use crate::hll::hash_set::HashSet;
+use crate::hll::list::List;
+use crate::hll::serialization::*;
+use crate::hll::{HllType, RESIZE_DENOM, RESIZE_NUMER, coupon};
+
+/// Current sketch mode
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+enum CurMode {
+    List = 0,
+    Set = 1,
+    Hll = 2,
+}
+
+#[derive(Debug, Clone)]
+pub struct HllSketch {
+    lg_config_k: u8,
+    mode: Mode,
+}
+
+impl PartialEq for HllSketch {
+    fn eq(&self, other: &Self) -> bool {
+        if self.lg_config_k != other.lg_config_k {
+            return false;
+        }
+
+        match (&self.mode, &other.mode) {
+            (
+                Mode::List {
+                    list: l1,
+                    hll_type: t1,
+                },
+                Mode::List {
+                    list: l2,
+                    hll_type: t2,
+                },
+            ) => l1 == l2 && t1 == t2,
+            (
+                Mode::Set {
+                    set: s1,
+                    hll_type: t1,
+                },
+                Mode::Set {
+                    set: s2,
+                    hll_type: t2,
+                },
+            ) => s1 == s2 && t1 == t2,
+            (Mode::Array4(a1), Mode::Array4(a2)) => a1 == a2,
+            (Mode::Array6(a1), Mode::Array6(a2)) => a1 == a2,
+            (Mode::Array8(a1), Mode::Array8(a2)) => a1 == a2,
+            _ => false, // Different modes are not equal
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+enum Mode {
+    List { list: List, hll_type: HllType },
+    Set { set: HashSet, hll_type: HllType },
+    Array4(Array4),
+    Array6(Array6),
+    Array8(Array8),
+}
+
+impl HllSketch {
+    /// Create a new HLL sketch
+    ///
+    /// # Arguments
+    /// * `lg_config_k` - Log2 of the number of buckets (K). Must be in [4, 21].
+    ///   - lg_k=4: 16 buckets, ~26% relative error
+    ///   - lg_k=12: 4096 buckets, ~1.6% relative error (common choice)
+    ///   - lg_k=21: 2M buckets, ~0.4% relative error
+    /// * `hll_type` - Target HLL array type (Hll4, Hll6, or Hll8)
+    ///
+    /// # Panics
+    /// Panics if lg_config_k is not in [4, 21]
+    pub fn new(lg_config_k: u8, hll_type: HllType) -> Self {
+        assert!(
+            lg_config_k > 4 && lg_config_k < 21,
+            "lg_config_k must be in [4, 21]"
+        );
+
+        let list = List::default();
+
+        Self {
+            lg_config_k,
+            mode: Mode::List { list, hll_type },
+        }
+    }
+
+    pub fn lg_config_k(&self) -> u8 {
+        self.lg_config_k
+    }
```

Comment:
```suggestion
    /// Return the configured lg_config_k.
    pub fn lg_config_k(&self) -> u8 {
        self.lg_config_k
    }
```

nit: we can later (or even now) turn on the `#![deny(missing_docs)]` to ensure all public APIs are well documented.